// @patternfly/react-table/dist/js/components/Table/utils/transformers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var selectable_1 = require("./decorators/selectable");
exports.selectable = selectable_1.selectable;
var sortable_1 = require("./decorators/sortable");
exports.sortable = sortable_1.sortable;
var cellActions_1 = require("./decorators/cellActions");
exports.cellActions = cellActions_1.cellActions;
var cellWidth_1 = require("./decorators/cellWidth");
exports.cellWidth = cellWidth_1.cellWidth;
var editable_1 = require("./decorators/editable");
exports.editable = editable_1.editable;
var wrappable_1 = require("./decorators/wrappable");
exports.breakWord = wrappable_1.breakWord;
exports.fitContent = wrappable_1.fitContent;
exports.nowrap = wrappable_1.nowrap;
exports.truncate = wrappable_1.truncate;
exports.wrappable = wrappable_1.wrappable;
var textCenter_1 = require("./decorators/textCenter");
exports.textCenter = textCenter_1.textCenter;
var collapsible_1 = require("./decorators/collapsible");
exports.collapsible = collapsible_1.collapsible;
exports.expandedRow = collapsible_1.expandedRow;
exports.expandable = collapsible_1.expandable;
var compoundExpand_1 = require("./decorators/compoundExpand");
exports.compoundExpand = compoundExpand_1.compoundExpand;
var headerCol_1 = require("./decorators/headerCol");
exports.headerCol = headerCol_1.headerCol;
var classNames_1 = require("./decorators/classNames");
exports.classNames = classNames_1.classNames;
exports.Visibility = classNames_1.Visibility;
const emptyTD = () => ({
    scope: '',
    component: 'td'
});
exports.emptyTD = emptyTD;
const scopeColTransformer = () => ({
    scope: 'col'
});
exports.scopeColTransformer = scopeColTransformer;
const emptyCol = (label) => (Object.assign({}, (label ? {} : { scope: '' })));
exports.emptyCol = emptyCol;
const parentId = (_value, { rowData }) => ({
    parentId: rowData.parent
});
exports.parentId = parentId;
const mapProps = (_label, { property, rowData }) => (Object.assign({}, (rowData[property] && rowData[property].props)));
exports.mapProps = mapProps;
